src/session/client-certificate.c

```diff
@@ -300,8 +368,39 @@ cockpit_session_client_certificate_map_user (const char *client_certificate_file
       return NULL;
     }
 
+  /* check if stdin is a Unix socket; then we are being called via cockpit-session@.service and need to
+   * check the cgroup of our peer */
+  struct ucred ucred;
+  int ws_pid_dirfd;
+  if (getsockopt (STDIN_FILENO, SOL_SOCKET, SO_PEERCRED, &ucred, &(socklen_t){sizeof ucred}) == 0)
+    {
+      debug ("unix socket mode, reading cgroup from peer pid %d", ucred.pid);
+      ws_pid_dirfd = open_proc_pid (ucred.pid);
+      int my_pid_dirfd = open_proc_pid (getpid ());
+      unsigned long ws_start_time = get_proc_pid_start_time (ws_pid_dirfd);
+      unsigned long my_start_time = get_proc_pid_start_time (my_pid_dirfd);
+      debug ("peer start time: %lu, my start time: %lu", ws_start_time, my_start_time);
+      close (my_pid_dirfd);
+
+      /* Guard against pid recycling: If a malicious user captures ws, keeps the socket in a forked child and exits
+       * the original pid, they can trap a different user to login, get the old pid (pointing ot their cgroup), and
+       * capture their session. To prevent that, require that ws must have started earlier than ourselves. */
+      if (my_start_time < ws_start_time)
+        {
+          warnx ("start time of this process (%lu) is older than cockpit-ws (%lu), pid recycling attack?",
+                  my_start_time, ws_start_time);
+          close (ws_pid_dirfd);
+          return NULL;
+        }
+    }
+  else {
+    ws_pid_dirfd = open_proc_pid (getpid ());
```

Comment:
If I understand correctly, this should only happen in a test case now, right?  When we run `cockpit-ws` as root and want it to directly spawn `cockpit-session`.

But: do we actually test certificate authentication in that test case?  It feels to me like this fallback is:
 - dead in production
 - maybe also dead in tests?

If so it would be nice to change this code to hard-bomb if `getsockopt()` fails.